ethcore/engines/authority-round/src/randomness.rs

```diff
@@ -0,0 +1,248 @@
+	PhaseConflict,
+	/// The smart contract reported that we already revealed something while still being in the
+	/// commit phase.
+	#[display(fmt = "Revealed during commit phase")]
+	RevealedInCommit,
+	/// Failed to load contract information.
+	#[display(fmt = "Error loading randomness contract information: {:?}", _0)]
+	LoadFailed(CallError),
+	/// Failed to load the stored encrypted random number.
+	#[display(fmt = "Failed to load random number from the randomness contract")]
+	BadRandNumber,
+	/// Failed to encrypt random number.
+	#[display(fmt = "Failed to encrypt random number: {}", _0)]
+	Crypto(CryptoError),
+	/// Failed to get the engine signer's public key.
+	#[display(fmt = "Failed to get the engine signer's public key")]
+	MissingPublicKey,
+}
+
+impl From<CryptoError> for PhaseError {
+	fn from(err: CryptoError) -> PhaseError {
+		PhaseError::Crypto(err)
+	}
+}
+
+impl RandomnessPhase {
+	/// Determine randomness generation state from the contract.
+	///
+	/// Calls various constant contract functions to determine the precise state that needs to be
+	/// handled (that is, the phase and whether or not the current validator still needs to send
+	/// commitments or reveal random numbers).
+	pub fn load(
+		contract: &BoundContract,
+		our_address: Address,
+	) -> Result<RandomnessPhase, PhaseError> {
+		// Determine the current round and which phase we are in.
+		let round = contract
+			.call_const(aura_random::functions::current_collect_round::call())
+			.map_err(PhaseError::LoadFailed)?;
+		let is_commit_phase = contract
+			.call_const(aura_random::functions::is_commit_phase::call())
+			.map_err(PhaseError::LoadFailed)?;
+
+		// Ensure we are not committing or revealing twice.
+		let committed = contract
+			.call_const(aura_random::functions::is_committed::call(
+				round,
+				our_address,
+			))
+			.map_err(PhaseError::LoadFailed)?;
+		let revealed: bool = contract
+			.call_const(aura_random::functions::sent_reveal::call(
+				round,
+				our_address,
+			))
+			.map_err(PhaseError::LoadFailed)?;
+
+		// With all the information known, we can determine the actual state we are in.
+		if is_commit_phase {
+			if revealed {
+				return Err(PhaseError::RevealedInCommit);
+			}
+
+			if !committed {
+				Ok(RandomnessPhase::BeforeCommit)
+			} else {
+				Ok(RandomnessPhase::Committed)
+			}
+		} else {
+			if !committed {
+				// We apparently entered too late to make a commitment, wait until we get a chance again.
+				return Ok(RandomnessPhase::Waiting);
+			}
+
+			if !revealed {
+				Ok(RandomnessPhase::Reveal { our_address, round })
+			} else {
+				Ok(RandomnessPhase::Waiting)
+			}
+		}
+	}
+
+	/// Advance the random seed construction process as far as possible.
+	///
+	/// Returns the encoded contract call necessary to advance the randomness contract's state.
+	///
+	/// **Warning**: The `advance()` function should be called only once per block state; otherwise
```

Comment:
Not sure what you mean by "block state"? Can you elaborate?

Reply:
I think the intent of this warning was to make sure we don't call `advance()` twice in the same block. I'll reword it.